include/charmlite/charmlite.hpp

```diff
@@ -1,21 +1,253 @@
-#ifndef __CMK_HH__
-#define __CMK_HH__
+#ifndef CHARMLITE_CHARMLITE_HPP
+#define CHARMLITE_CHARMLITE_HPP
 
-#include <charmlite/core/collection.hpp>
 #include <charmlite/core/core.hpp>
-#include <charmlite/core/proxy.hpp>
+
 #include <charmlite/algorithms/reduction.hpp>
 
-// ( no reordering )
-#include <charmlite/core/charmlite.impl.hpp>
+#include <charmlite/utilities/math.hpp>
+#include <charmlite/utilities/traits.hpp>
 
 namespace cmk {
+
+    template <typename Message, typename... Args>
+    inline message_ptr<Message> make_message(Args&&... args)
+    {
+        auto* msg = new Message(std::forward<Args>(args)...);
+        return message_ptr<Message>(msg);
+    }
+
+    void start_fn_(int, char**);
+
+    inline collection_base_* lookup(collection_index_t idx)
+    {
+        auto& tab = CpvAccess(collection_table_);
+        auto find = tab.find(idx);
+        if (find == std::end(tab))
+        {
+            return nullptr;
+        }
+        else
+        {
+            return (find->second).get();
+        }
+    }
+
+    inline void initialize(int argc, char** argv)
+    {
+        ConverseInit(argc, argv, start_fn_, 1, 1);
+        initialize_globals_();
+#if CHARMLITE_TOPOLOGY
+        TopoManager_init();
+        CmiNodeAllBarrier();
+        CmiInitCPUAffinity(argv);
+        CmiInitMemAffinity(argv);
+        CmiInitCPUTopology(argv);
+        TopoManager_reset();    // initialize TopoManager singleton
+        _topoTree = ST_RecursivePartition_getTreeInfo(0);
+#endif
+        CmiNodeAllBarrier();
+    }
+
+    inline void send(message_ptr<>&& msg, bool immediate)
+    {
+        if (immediate)
+        {
+            converse_handler_(msg.release());
+        }
+        else
+        {
+            send(std::move(msg));
+        }
+    }
+
+    inline void finalize(void)
+    {
+        CsdScheduleForever();
+        ConverseExit();
+    }
+
+    void exit(message_ptr<>&& msg);
+
+    template <entry_fn_t Fn, bool Constructor>
+    static entry_id_t register_entry_fn_(void)
```

Comment:
Perhaps we could move all the `register_` functions (and their helpers) into _registration.hpp_?

Reply:
Yeah, that's actually a better thing to do.